sigmf/archive.py

```diff
@@ -0,0 +1,140 @@
+                   if specified will be used to determine the directory and
+                   file names within the archive. `fileobj` won't be closed.
+                   For example: if `name` == "archive1" and fileobj is given,
+                   a tar archive will be written to fileobj with the
+                   following structure:
+                       - archive1/
+                         - archive1.sigmf-meta
+                         - archive1.sigmf-data
+
+    """
+    def __init__(self, sigmffile, name=None, fileobj=None):
+        self.sigmffile = sigmffile
+        self.name = name
+        self.fileobj = fileobj
+
+        self._validate()
+
+        archive_name = self._get_name()
+        sigmf_fd = self._open()
+        sigmf_archive = tarfile.TarFile(mode="w",
+                                        fileobj=sigmf_fd)
+        tmpdir = tempfile.mkdtemp()
+        sigmf_md_filename = archive_name + SIGMF_METADATA_EXT
+        sigmf_md_path = os.path.join(tmpdir, sigmf_md_filename)
+        sigmf_data_filename = archive_name + SIGMF_DATASET_EXT
+        sigmf_data_path = os.path.join(tmpdir, sigmf_data_filename)
+
+        with open(sigmf_md_path, "w") as mdfile:
+            self.sigmffile.dump(mdfile, pretty=True)
+
+        shutil.copy(self.sigmffile.data_file, sigmf_data_path)
+
+        def chmod(tarinfo):
+            if tarinfo.isdir():
+                tarinfo.mode = 0o755  # dwrxw-rw-r
+            else:
+                tarinfo.mode = 0o644  # -wr-r--r--
+            return tarinfo
+
+        sigmf_archive.add(tmpdir, arcname=archive_name, filter=chmod)
+        sigmf_archive.close()
+        if not fileobj:
+            sigmf_fd.close()
+
+        shutil.rmtree(tmpdir)
+
+        self.path = sigmf_archive.name
+
+    def _validate(self):
+        name = self.name
+        if name and "." in name and not name.endswith(SIGMF_ARCHIVE_EXT):
+            apparent_ext = os.path.splitext(name)[-1]
+            err = "extension {} != {}".format(apparent_ext, SIGMF_ARCHIVE_EXT)
+            raise error.SigMFFileError(err)
+
+        valid_md = self.sigmffile.validate()
+        if not valid_md:
+            err = "invalid metadata - {!s}"
+            raise error.SigMFValidationError(err.format(valid_md))
+
+        if not self.sigmffile.data_file:
+            err = "no data file - use `set_data_file`"
+            raise error.SigMFFileError(err)
+
+    def _get_name(self):
+        if self.fileobj and not self.name:
+            filename = os.path.split(self.fileobj.name)[-1]
+        else:
+            filename = os.path.split(self.name)[-1]
+
+        archive_name = os.path.splitext(filename)[0]
+        return archive_name
+
+    def _open(self):
+        if self.fileobj:
+            sigmf_fd = self.fileobj
+        else:
+            has_ext = self.name.endswith(SIGMF_ARCHIVE_EXT)
+            name = self.name if has_ext else self.name + SIGMF_ARCHIVE_EXT
+            sigmf_fd = open(name, "wb")
```

Comment:
I think you have a good point, and I would prefer that we don't create new exception classes for the sole purpose of renaming them to `SigMF` exceptions.

My one concern with the above, though, is that if this code is being used more indirectly by a user, it may not be clear to them **which** file was not writable, and the Python3 traceback isn't immediately useful for identifying which filepath was the problem.

So, how about passing on the built-in exception, but printing a bit more information about which file the user needs to fix?